relay-server/src/actors/outcome.rs

```
@@ -622,31 +704,144 @@ impl Actor for HttpOutcomeProducer {
     type Context = Context<Self>;
 }
 
-impl Supervised for HttpOutcomeProducer {}
+impl Handler<TrackRawOutcome> for HttpOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
 
-impl SystemService for HttpOutcomeProducer {}
+    fn handle(&mut self, msg: TrackRawOutcome, ctx: &mut Self::Context) -> Self::Result {
+        self.send_http_message(msg, ctx);
+        Ok(())
+    }
+}
 
-impl Default for HttpOutcomeProducer {
+impl Handler<TrackOutcome> for HttpOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
+
+    fn handle(&mut self, msg: TrackOutcome, ctx: &mut Self::Context) -> Self::Result {
+        self.handle(TrackRawOutcome::from_outcome(msg, &self.config), ctx)
+    }
+}
+
+struct ClientReportOutcomeProducer {}
+
+impl Actor for ClientReportOutcomeProducer {
+    type Context = Context<Self>;
+}
+
+impl Handler<TrackOutcome> for ClientReportOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
+
+    fn handle(&mut self, msg: TrackOutcome, _ctx: &mut Self::Context) -> Self::Result {
+        let mut client_report = ClientReport {
+            timestamp: Some(UnixTimestamp::from_secs(
+                msg.timestamp.timestamp().try_into().unwrap_or(0),
+            )),
+            ..Default::default()
+        };
+
+        // The outcome type determines what field to place the outcome in:
+        let discarded_events = match msg.outcome {
+            Outcome::Filtered(_) => &mut client_report._server_filtered_events,
+            Outcome::FilteredSampling(_) => &mut client_report._server_filtered_sampling_events,
+            Outcome::RateLimited(_) => &mut client_report._server_rate_limited_events,
+            _ => {
+                // Cannot convert this outcome to a client report.
+                return Ok(());
```

Note that only these three outcome types will be emitted. Everything else is silently dropped.